load asset value config

Nintorac opened the pull request:
### Summary & Motivation

Allowed supply of configuration to `load_asset_value` and propagated information to where it's needed. 

fixes #10990

### How I Tested These Changes

not as yet

dpeng817 commented on python_modules/dagster/dagster/_core/definitions/repository_definition.py:
```diff
@@ -1403,6 +1403,7 @@ def load_asset_value(
         python_type: Optional[Type] = None,
         instance: Optional[DagsterInstance] = None,
         partition_key: Optional[str] = None,
+        config: Optional[Any] = None
```
would want to see an entry in the doc string

dpeng817 commented on python_modules/dagster/dagster/_core/storage/asset_value_loader.py:
```diff
@@ -89,13 +91,15 @@ def load_asset_value(
         io_manager_def = resource_defs[io_manager_key]
         required_resource_keys = io_manager_def.required_resource_keys | {io_manager_key}
-
+        config = config or {}
+        resource_config = config.get('resources', {})
```
it seems like the only thing we're pulling off of the config here is the resources config. I wonder if we would be better served by just making the parameter called `resources_config` directly, and allowing you to configure each resource by key without having to specify the top level `resources` key to the config dictionary.

Nintorac replied:
I am passing the config into `build_input_context`, is it not required there?

Nintorac replied:
As well, I tend to pass the same config around with config for the whole project, so it's a little less boiler plate and simplifies things a bit

dpeng817 replied:
The config argument on `build_input_context` is for input config specifically, so the config argument provided here wouldn't match up. `build_input_context` also has a `resource_config` argument (should probably be called resources_config) that allows for providing config for the resources.

Regarding passing the same config around, that's definitely reasonable and I imagine many users do the same thing, but it seems like the difference is just between passing `run_config` and `run_config.get("resources")`, no? In which case, I'm inclined to scope it down to just resources_config to keep the data model simpler.

Nintorac replied:
OK sounds good, I have updated accordingly

dpeng817 reviewed (changes requested):
Some things to consider. Definitely also going to want to add tests; the common case where the io manager needs input config, but then also the case where a resource required by the IO manager needs config.

Nintorac commented:
I think this is ready now, updates include:
- added a way to pass resource_config into asset value loader
- added tests on asset_value_loader for iomanagers with config
- added tests on asset_value_loader for iomanagers with resources with config
- added tests on repository load_asset_value for iomanagers with config
- added tests on repository load_asset_value for iomanagers with resources with config

dpeng817 commented on python_modules/dagster/dagster/_core/storage/asset_value_loader.py:
```diff
@@ -75,11 +81,14 @@ def load_asset_value(
                 be returned inside `load_input` by `context.dagster_type.typing_type`.
             partition_key (Optional[str]): The partition of the asset to load.
+            resource_config (Optional[Any]): A dictionary of resource configurations to be passed
+                to the :py:class:`IOManager`.
```
nit: would be more accurate to `resource configurations to be passed to resources required by the IOManager`

sryza replied:
it's to the IOManager too though, right?

dpeng817 replied:
that is also true, yea

dpeng817 reviewed (commented):
https://buildkite.com/dagster/dagster/builds/43903 - this is looking super close - will follow up with results of this build!

dpeng817 commented on python_modules/dagster/dagster_tests/storage_tests/test_asset_value_loader.py:
```diff
@@ -216,3 +216,66 @@ def repo():
+            return context.resources.io_resource
+
+    @io_manager(required_resource_keys={"io_resource"})
+    def my_io_manager(context):
```
```suggestion
    def my_io_manager():
```
Linter complains about this

dpeng817 reviewed (approved):
Just that one lint error and this will be good to go

Nintorac commented:
Sorry for the delay -- Ruff looks cool. A fair bit faster!

This should be ready now :)
